### ISSUE TITLE
centralized value conversion

### ISSUE THREAD

**Issue by sanderr** [inmanta/inmanta#1782]
The compiler needs to do a couple of value conversions:
- parsing: Inmanta -> internal
- plugin call: internal -> Python
- plugin return: Python -> internal
- export: internal -> JSON

I believe it might be benificial to create a value transformer framework that can be called on to do value conversions from any representation to any other. For many values this transformation will be just the identity function, but `None` values for example have different representations in three of four cases: `null`, `NoneValue()`, `None`.

A possible format for this framework could be
```
class ValueTransformer:
    def __init__(self, source: ValueRepresentation, dest: ValueRepresentation) -> None:
        self.source: ValueRepresentation = source
        self.dest: ValueRepresentation = dest

    def transform(value: object) -> object:
        return self.dest.from_python(self.source.to_python())
```

This can probably be improved, but it gives a general idea. The advantage of such a module would mainly be consistency in value translation. The relevance of this consistency is shown by the existence of issues such as #817 and #674.

**Comment by sanderr** [comment]
This might be related as well
https://github.com/inmanta/inmanta/blob/958b0f5539b4c56a59445a0b544bc5be1b1424f2/src/inmanta/execute/proxy.py#L78

**Comment by wouterdb** [comment]
Interesting notion, good suggestion.

first remarks
- I would not use python objects as intermediate representation, but rather the compiler internal represenation
- plugin call input has the special property that it is a proxy: it lazily perform the translation when the model is navigated.
- parsing is also special, as it only produces literal values and the source representation is not directly typed, but types are derived from the lexical/parser context
- export has model exporter, types exporter, resource exporter, I assume this is about the model exporter?

What would the transformer interface look like? 
What facilities would it offer?

**Comment by sanderr** [comment]
- I agree parsing is somewhat special, though I also believe it would be beneficial to integrate parts of it in the transformer. For example, currently the type "number[]?" is interpreted by the parser and the plugin module separately, which could lead to inconsistencies.
- I was indeed talking about the model exporter

I'm not sure yet what the interface would look like, this would require some further investigation. I believe it should offer transformations from any type system to any other type system, for values as well as the types themselves. This could possibly integrate with value casts as well, though I'm not sure if that really belongs here.

**Comment by sanderr** [comment]
Also value casts in `inmanta.ast.type` could benefit a lot from this

**Comment by wouterdb** [comment]
I would like to see the design a bit more clear. Can you update the ticket with the exact capabilities you envision and How this would look?

- Interpretation of type modifiers (like string[]?) would indeed be useful. I don't know if it fits a conversion framework, as it is parsing. (I consider conversion something that starts from a structured object and produces something out of it, Parsing is consider something that starts from a string and produces a structured object.)
- For the conversion, would it be like a visitor on the types?

**Comment by sanderr** [comment]
I will think on a more detailed proposal. For now, I don't entirely agree with your first point. Parsing is indeed something that starts from a string and produces a structured object, but that string is itself a structured object. Interpreting types as a parsing atom and delegating the conversion from the structured string representation to a structured internal representation seems like a worthwhile trade-off considering the abstraction gains given that this type interpretation is distributed over the whole compiler, not only in the parser.

**Comment by wouterdb** [comment]
I eagerly await the proposall

**Comment by sanderr** [comment]
I think what I had in mind might be a bit too grand for the requirements. I've scratched my more ambitious ideas and reanalyzed the problem:

- the model export is different since it does a lot more than type or value conversion.
- there are essentially three domains: inmanta (inmanta types, inmanta values) -- internal (internal types, internal values) -- plugin (inmanta types, Python values). Where inmanta types and values are strings.
- conversion between inmanta and internal domain is unidirectional (parsing)
- value conversion between internal and plugin domain is bidirectional but simple, only None values have a different representation.
- type conversion between internal and plugin domain is unidirectional: inmanta types get converted to internal types.

Conclusions:
- a centralized value conversion is probably not worth the effort. It would probably lead to a cleaner design but at this point the chance of unintentially breaking things is low since only None values need converting and it all happens at dedicated parts of the code.
- centralized type conversion from inmanta strings to internal is the core of why I created this ticket. Changes in the internal type system might require updates to this conversion. Since it's currently distributed over at least two places (plugins and parsing) inconsistensies may arise. **This is the part we will try to solve.**
- In the future some differences might be introduced between inmanta types and plugin type annotations (e.g. lists of entities). The type conversion should account for this.

Semi-concrete proposal:
- centralize type interpretation: inmanta string -> internal. Something like `Plugin.to_type()` in `inmanta.ast.type`. It returns either `Optional[List[Number]]` (option 1) or a list of transformations that need to be applied to the primitive `("number", [List, Optional])` to seperate type transformations from type lookups in the namespace (option 2).
- parser delegates type interpretation to this function (it recognizes that `number[]?` is a type but then passes it on as a whole to the interpreter)
- For plugins: type interpreter that inherits from the default one. This might not even be required if option 2 is chosen, since then the plugin can do it's own interpretation before passing the type on to the resolver. Whichever type interpreter is chosen, the current `Plugin.to_type()` method calls / is replaced by a call to the interpreter.

Evaluation:
1. Are my assumptions (the first bullet list) correct?
2. Are my conclusions correct?
3. Is the proposal clear?
4. Does the proposal solve the problem?
5. Is the proposal sufficiently concrete for this phase of the design?

**Comment by sanderr** [comment]
Type interpretation is also very relevant for LSM. Internal types get stringified, then interpreted by LSM to convert them to pydantic types. LSM issue 75 illustrates the relevance.

**Comment by wouterdb** [comment]
Notes: 
 - value conversion between internal and plugin domain is not that simple: see DynamicProxy

**Comment by sanderr** [comment]
I'll convert this to a PR in the inmanta-telco repo once I've completed my current issues in progress.

On your notes:
- You're right, I wasn't really aware of those up to this point.
- True, an attempt should be made to take that in as well.
- Good catch, I mistyped, it should be `Nullable[TypedList[Number]]` which is the internal representation of the type.
- Currently, the parser interprets multi and nullable, and the base type is later looked up in the namespace. The other place is in `Plugin.get_type()` which does something very similar. The proposed new location would be in `inmanta.ast.type`.
